### PR TITLE
Summary quality (grammar & meaning) classifier using Bert and a FeedForward network

### PR DESCRIPTION
**Pull request by yechenh**

### CODE REVIEW

File: classifier/README.md [hunk 1]
```diff
@@ -1 +1,57 @@
+# Summary Quality Classifier
+
+This classifier assesses the quality of text summarization result in terms 
+of grammar and meaning preservation. This model modified from the
```
Comment (XiangZeng2020) [inline classifier/README.md:4]: s/modified/is modified

File: classifier/README.md [hunk 2]
```diff
@@ -1 +1,57 @@
+# Summary Quality Classifier
+
+This classifier assesses the quality of text summarization result in terms 
+of grammar and meaning preservation. This model modified from the 
+[LaserTagger](https://github.com/google-research/lasertagger) script, and 
+is built on Python 3, Tensorflow and 
+[BERT](https://github.com/google-research/bert). 
+
+BERT transformer model is used as the encoder, and for the decoder, a single
+feed-forward pass is applied to the decoder. This model should be trained on
```
Comment (XiangZeng2020) [inline classifier/README.md:10]: s/feed-forward pass/feed-forward network

File: classifier/README.md [hunk 3]
```diff
@@ -1 +1,57 @@
...
+the 
+[Microsoft Abstractive Text Compression Dataset](https://www.microsoft.com/en-us/download/confirmation.aspx?id=54262).
+This dataset contains ratings for each source-summary pair, and is scored on
+grammar and meaning on a 0-2 scale.
```
Comment (XiangZeng2020) [inline classifier/README.md:14]: why 0-2 not 0-1 scale?

File: classifier/README.md [hunk 4]
```diff
@@ -1 +1,57 @@
...
+is needed for an experiment with this model.
+
+### 1. Preprocessing
+
+Download the 
+[Microsoft Abstractive Text Compression Dataset](https://www.microsoft.com/en-us/download/confirmation.aspx?id=54262)
+and unzip. You should find a folder named "RawData". Run the following to 
+preprocess the dataset. 
+
+```
```
Comment (XiangZeng2020) [inline classifier/README.md:28]:
```script

File: classifier/README.md [hunk 5]
```diff
@@ -1 +1,57 @@
...
+export NUM_TUNE=500
+export NUM_VALID=500
+export INPUT_DIR=/path/to/RawData
+
+python preprocess_MS_dataset_for_classifier.py INPUT_DIR NUM_TUNE NUM_VALID
+```
+
+Six tsv files will be saved at your home directory. Three corresponding to grammaring scoring,
```
Comment (XiangZeng2020) [inline classifier/README.md:36]: s/grammaring/grammar
s/corresponding/correspond

File: classifier/README.md [hunk 6]
```diff
@@ -1 +1,57 @@
...
+and three for meaning preservation scoring. Next, run the following command
```
Comment (XiangZeng2020) [inline classifier/README.md:37]: s/for/is for

File: classifier/utils.py [hunk 7]
```diff
@@ -40,77 +40,31 @@ def get_token_list(text):
...
-    input_format):
+def yield_sources_and_targets_meaning(input_file):
   """Reads and yields source lists and targets from the input file.
 
   Args:
     input_file: Path to the input file.
-    input_format: Format of the input file.
 
   Yields:
     Tuple with (list of source texts, target text).
```
Comment (XiangZeng2020) [inline classifier/utils.py:50]: A tuple of (list of source texts, target text, score).

File: classifier/utils.py [hunk 8]
```diff
@@ -40,77 +40,31 @@ def get_token_list(text):
...
+        
+        
+def yield_sources_and_targets_grammar(input_file):
+  """Reads and yields source lists and targets from the input file.
 
   Args:
     input_file: Path to the input file.
+
+  Yields:
+    Tuple with (list of source texts, target text).
```
Comment (XiangZeng2020) [inline classifier/utils.py:65]: A tuple of (list of source texts, target text, score)

File: classifier/run_classifier_utils.py [hunk 9]
```diff
@@ -85,68 +61,74 @@ def __init__(self, config, num_tags,
...
-      # Size of the output vocabulary which contains the tags + begin and end
-      # tokens used by the Transformer decoder.
-      output_vocab_size = self._num_tags + 2
-      params = _get_decoder_params(self._config, self._use_tpu,
-                                   self._max_seq_length, output_vocab_size)
-      decoder = transformer_decoder.TransformerDecoder(params, is_training)
-      logits = decoder(input_mask, final_hidden, labels)
+    final_hidden_source = model_source.get_sequence_output()
+    
+    if self._classifier_type == "Meaning":
```
Comment (XiangZeng2020) [inline classifier/run_classifier_utils.py:91]: QQ: I'm wondering why you choose to train two BERT models and concatenate them here?

Comment (yechenh) [reply classifier/run_classifier_utils.py:91]: Bert learns the context of the input text, and I do not want the context of either the source or the summary to affect the embedding of the other. This is my reasoning for this decision. What would you suggest doing here? Thanks!

Comment (XiangZeng2020) [reply classifier/run_classifier_utils.py:91]: Are input_ids_source and input_ids_summary aligned? (I think so since you can concatenate them in the output)
If so usually people concatenate their embedding first then feed into the BERT model.

Comment (yechenh) [reply classifier/run_classifier_utils.py:91]: I'm not sure what you mean by aligned? They are one pair of source and summary, but they are of different length usually. They are padded to the same length and then fed into BERT separately. If I directly concatenate them, I was worried that they would be treated like one continuous piece of text. Particularly, from what I read online, the first token in the output of BERT kind of summarizes the meaning of the input text. If I feed source and summary separately, they would each get such a token in the embedding, which would be kind of useful in this case because we hope that the classifier can learn how well the meaning of the summary matches that of the source. But if I concatenate them, then there would only be one such token representing the meaning of both source and summary, and I am not sure whether that would impact the performance.

Comment (XiangZeng2020) [reply classifier/run_classifier_utils.py:91]: I see, you get the first output of the BERT for both. Then it makes sense to me. Thanks for the explanation!

File: classifier/README.md (hunk 3 above)
Comment (yechenh) [reply classifier/README.md:14]: The dataset has three levels of scoring for grammar and for meaning. I am not sure why they did this. I converted the scoring into 0, 1, 2 in preprocessing.

File: classifier/README.md (hunk 1 above)
Comment (yechenh) [reply classifier/README.md]: Done

File: classifier/README.md (hunk 2 above)
Comment (yechenh) [reply classifier/README.md]: Done

File: classifier/README.md (hunk 5 above)
Comment (yechenh) [reply classifier/README.md]: Done

File: classifier/README.md (hunk 6 above)
Comment (yechenh) [reply classifier/README.md]: Done

File: classifier/utils.py (hunk 7 above)
Comment (yechenh) [reply classifier/utils.py]: Done

File: classifier/utils.py (hunk 8 above)
Comment (yechenh) [reply classifier/utils.py]: Done

File: classifier/README.md [hunk 10]
```diff
@@ -30,16 +32,38 @@ export NUM_TUNE=500
...
 
 Six tsv files will be saved at your home directory. Three correspond to grammar scoring,
-and three are for meaning preservation scoring. Next, run the following command 
-on all six tsv files to convert the samples to tf_record (i.e. change the `INPUT_FILE`
-to the path to the training, tuning, and validation files, change `TYPE_OF_INPUT` to 
-grammar/meaning_train/tune/valid, and change `classifier_type` to either 
-Grammar or Meaning respectively).
+and three are for meaning preservation scoring. 
 
+Note that only 14% of the Microsoft dataset are unacceptable grammar or meaning samples (negative samples). To
```
Comment (XiangZeng2020) [inline classifier/README.md:41]: meaning -> meaningless?

File: classifier/README.md [hunk 11]
```diff
@@ -30,16 +32,38 @@ export NUM_TUNE=500
...
+can be used. Run the following command to preprocess this dataset and add all its negative samples to the training set for 
+the grammar classifier.
+ ```
+ python preprocess_cola_dataset_for_classifier.py /path/to/CoLA/data /path/to/grammar/training/tsv
+ ```
+A tsv file named classifier_mixed_training_set_grammar.tsv will be saved, which mixes the Microsoft dataset
+with the negative samples from the CoLA dataset. 
+
+To mix the training datasets with other synthetic negative samples to achieve a desirable 
+negative sample ratio, you can use
```
Comment (XiangZeng2020) [inline classifier/README.md:53]: s/use/run

File: classifier/README.md [hunk 12]
```diff
@@ -68,33 +92,50 @@ python run_classifier.py \
...
  --vocab_file=${BERT_BASE_DIR}/vocab.txt
  --saved_model=path/to/exported/model \
  --batch_size=500 \
  --classifier_type=Grammar
 ```
-To make Meaning scoring inferences, change the classifier_type to Meaning.
+To make meaning scoring inferences, change the classifier_type to Meaning.
+
+Compute the accuracy, precision, and recall score of the predictions:
+```
```
Comment (XiangZeng2020) [inline classifier/README.md:127]:
```script

File: classifier/README.md [hunk 13]
```diff
@@ -68,33 +92,50 @@ python run_classifier.py \
...
+python score_main.py path/to/output/prediction/file
+```
+Example output:
+```
```
Comment (XiangZeng2020) [inline classifier/README.md:131]: you need to add an empty line for the block to work

File: classifier/mixing_in_negative_samples.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        new_lines = f.readlines()
+    
+    if number_of_new_lines < 0:
+        raise ValueError("The target ratio is too small.")
+    if number_of_new_lines > len(new_lines):
+        raise ValueError("The target ratio is too big given the number of synthetic data.")
+    
+    random.shuffle(new_lines)
+    shuffled_index = list(range(len(original_lines) + number_of_new_lines))
+    random.shuffle(shuffled_index)
```
Comment (XiangZeng2020) [inline classifier/mixing_in_negative_samples.py:49]: can you comment what you are doing here? it's not very obvious.

File: classifier/mixing_in_negative_samples.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    
+    # Grammar classifier only needs one input while meaning classifier needs two
+    number_of_inputs = len(original_lines[0].strip().split("\t")) - 1
+    
+    with open(os.path.expanduser(output_file_path), 'wt') as out_file:
+        tsv_writer = csv.writer(out_file, delimiter='\t')        
+        for index in shuffled_index:
```
Comment (XiangZeng2020) [inline classifier/mixing_in_negative_samples.py:56]: reduce indent level

File: classifier/mixing_in_negative_samples.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+      target_ratio: the target ratio of negative samples
+    
+    Returns:
+      The number of negative samples needed to achieve the target ratio.
+    """
+    total_number_of_samples = 0
+    original_negative_sample_count = 0
+    with open(original_data_file) as tsv_file:
+        read_tsv = csv.reader(tsv_file, delimiter="\t")
+        for line in read_tsv:
```
Comment (XiangZeng2020) [inline classifier/mixing_in_negative_samples.py:81]: reduce indent level

File: classifier/preprocess_cola_dataset_for_classifier.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import argparse
+import csv
+import os
+import random
+
```
Comment (XiangZeng2020) [inline classifier/preprocess_cola_dataset_for_classifier.py:18]: delete extra empty line

File: classifier/preprocess_cola_dataset_for_classifier.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+import numpy as np
+
+import preprocess_utils
+""" Preprocess the CoLA (The Corpus of Linguistic Acceptability) grammar
```
Comment (XiangZeng2020) [inline classifier/preprocess_cola_dataset_for_classifier.py:22]: move file comment before the imports.

File: classifier/preprocess_cola_dataset_for_classifier.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    data_file = os.path.expanduser(args.raw_data_file)
+    if not os.path.isfile(data_file):
+        raise Exception("Data file not found.")
+    
+    sentences_positive = []
+    sentences_negative = []
+    
+    with open(data_file) as tsv_file:
+        read_tsv = csv.reader(tsv_file, delimiter="\t")
+        for line in read_tsv:
```
Comment (XiangZeng2020) [inline classifier/preprocess_cola_dataset_for_classifier.py:43]: reduce indent level

File: classifier/README.md (hunk 10 above)
Comment (yechenh) [reply classifier/README.md]: Done.

File: classifier/README.md (hunk 11 above)
Comment (yechenh) [reply classifier/README.md]: Done

File: classifier/README.md (hunk 12 above)
Comment (yechenh) [reply classifier/README.md:128]: Done

File: classifier/README.md (hunk 13 above)
Comment (yechenh) [reply classifier/README.md:133]: Done

File: classifier/mixing_in_negative_samples.py (hunk 14 above)
Comment (yechenh) [reply classifier/mixing_in_negative_samples.py:50]: I am reshuffling the dataset after adding new samples (see " for index in shuffled_index: " on line 57)

File: classifier/mixing_in_negative_samples.py (hunk 15 above)
Comment (yechenh) [reply classifier/mixing_in_negative_samples.py]: Done

File: classifier/mixing_in_negative_samples.py (hunk 16 above)
Comment (yechenh) [reply classifier/mixing_in_negative_samples.py]: Done

File: classifier/preprocess_cola_dataset_for_classifier.py (hunk 19 above)
Comment (yechenh) [reply classifier/preprocess_cola_dataset_for_classifier.py]: Done

File: classifier/preprocess_cola_dataset_for_classifier.py (hunk 18 above)
Comment (yechenh) [reply classifier/preprocess_cola_dataset_for_classifier.py]: Done.

File: classifier/preprocess_cola_dataset_for_classifier.py [hunk 20]
```diff
@@ -38,14 +38,15 @@ def main(args):
...
-
+    tsv_file = open(data_file)
+    read_tsv = csv.reader(tsv_file, delimiter="\t")
+    for line in read_tsv:
+        if int(line[1]) == 1:
+            sentences_positive.append(line[3])
+        else:
+            sentences_negative.append(line[3])
+    tsv_file.close()
+
```
Comment (XiangZeng2020) [inline classifier/preprocess_cola_dataset_for_classifier.py:49]: Sorry about the confusion in my previous comment. 
The original with comment should be kept,
I just meant you can make the for loop be aligned to the with:

with open(data_file) as tsv_file:
        read_tsv = csv.reader(tsv_file, delimiter="\t")
for line in read_tsv:
        if int(line[1]) == 1:
            sentences_positive.append(line[3])
        else:
            sentences_negative.append(line[3])

Comment (yechenh) [reply classifier/preprocess_cola_dataset_for_classifier.py:49]: I don't think I can use read_tsv (like for line in read_tsv) outside the with open as it will throw  "I/O operation on closed file." error. So if I keep the with open structure, I guess I will have to keep the for loop inside the with open. Which structure would you recommend? Thanks!

Comment (XiangZeng2020) [reply classifier/preprocess_cola_dataset_for_classifier.py:49]: you can just keep your original version. Sorry I didn't realize read_tsv is used later.

Comment (yechenh) [reply classifier/preprocess_cola_dataset_for_classifier.py]: Ok! Thanks!
